### ISSUE TITLE
Iterating "Big" (12k doc) Cursors

### ISSUE THREAD

**Issue by XescuGC** [arangodb/arangojs#59]
I have a question about how to iterate "big" cursors, reading the [arangojs](https://github.com/arangodb/arangojs#cursor-api) documentation and the [ArangoDB](https://docs.arangodb.com/HttpAqlQueryCursor/QueryResults.html) documentation, i've found that the cursors are fetched via batches: 

But i think i'm doing something wrong with it:

```javascript
db.query(AQB.for('page').in('pages').return('page'), (err, cursor) => {
  while ( cursor.hasNext() ) {
    cursor.next( (err, val) => {
      // use val to throw it to a queue
    }
  }
}
```

It only fetches the first 1k of the 12k documents.

I've readed the [code](https://github.com/arangodb/arangojs/blob/master/lib/cursor.js#L69) and i've seen that the `#next` uses the `#_more` wich fetches the next cursor. I think i'm doing something wrong but i don't know exactly what

**Comment by pluma** [comment]
I think the REST API limits query results to 1k by default independent of the batchSize. I'll have to dig into it and see if the driver misbehaves here or how the limit can be disabled.

**Comment by jsteemann** [comment]
Just to clarify what the server-side API for cursors does/expects:

The REST API for cursors has a `batchSize` attribute, which currently defaults to 1000. The initial call to POST `/_api/cursor` will return the amount specified in `batchSize` (i.e. 1000 by default), and further cursor results need to be fetched by the client via subsequent calls to PUT `/_api/cursor/<cursorId>` (cursorId will be returned in the response to the first call). This way the overall cursor result can be bigger than `batchSize`, but the client driver may have to issue multiple requests to fetch it all.

**Comment by pluma** [comment]
@jsteemann Indeed. The driver *should* be performing additional requests if necessary and if `hasMore` is true. The batchSize should have no effect on this. I originally tested this by setting the batchSize to a very small value. I suspect this is a regression in the driver rather than an API issue.

**Comment by jsteemann** [comment]
@pluma: thanks for the clarification!

**Comment by XescuGC** [comment]
@jsteemann yes that's the way it should work as specified on the ArangoDB documentation, by the way, the code example i've shown before, when the cursor reaches the 1k docs, the `#hasNext()` returns true because the `#_hasMores` is true.

**Comment by XescuGC** [comment]
What i've see debugging the lib is that the `#_more` is called, i put a semafor because of the `while` was called infinite times (so it fetches the API infinite times), and the result is that it calls the server API but the server never answers it, it never calls the [callback](https://github.com/arangodb/arangojs/blob/c83d7d8127f3eaf1dc58499fd89caa587192bef4/lib/cursor.js#L33). Which is the method that you use to fetch all the docs of a cursor? Because the `#next` is asynchronous so the `while` goes one without caring the result of `#next()` until the `#hasNext()` is false.
If you want me to try something say it because for me it's easy to test it jeje.

**Comment by pluma** [comment]
Okay, I think the problem is in how you're using ArangoJS.

While `hasNext` is synchronous, `next` is asynchronous (because it may need to peform another request when the local results have been depleted). Putting it in a synchronous loop will simply break things.

What you should be doing is something like this:

```js
db.query(AQB.for('page').in('pages').return('page'), (err, cursor) => {
  cursor.each(val => {
      // use val to throw it to a queue
  });
}
```

The iteration methods ("each", "every", etc) fetch additional data behind the scenes whenever necessary and will iterate over all results until certain conditions are met (e.g. "each" and "every" can be aborted by explicitly returning `false` and "some" can be aborted by returning `true`)

**Comment by pluma** [comment]
@XescuGC Can you verify the code works using `cursor.each`?

**Comment by pluma** [comment]
I'm unable to reproduce this issue in 4.0.0-alpha3 when using each.

**Comment by XescuGC** [comment]
Well now i get a wierd error:

```
FATAL ERROR: CALL_AND_RETRY_2 Allocation failed - process out of memory
Abort trap: 6
```
What i can see is that the lib first fetch (`#_drain`) all docs, the 12k, and then iterates over them? Well my docs are big ones and I think it can't load them to memory.

Do i try it with the 4.0.0-alpha3?

**Comment by XescuGC** [comment]
If in the query i only return one attribute of the `page` it works :smile:, otherwise it breaks with the error before.

**Comment by jsteemann** [comment]
`FATAL ERROR: CALL_AND_RETRY_2 Allocation failed - process out of memory` is a fatal error thrown by V8. When thrown, it will terminate execution and kill node.js. It may help to request only the actually desired document attributes instead of the full documents if the documents are really big. 

You can use an AQL query for that, i.e. `FOR page IN pages RETURN page.name` to return just the `name` attribute for each document in `pages`. The same would apply for `SELECT * FROM pages` vs. `SELECT name FROM pages` when using SQL and trying to reduce the result set sizes.

**Comment by XescuGC** [comment]
Yes, this is what i did: `AQB.for('page').in('pages').return('page.url')`, but i think that the behavior of the `#each`, which i think it's to consume all the cursor and save it to the `#_result` and then iterate it, it's not the correct one, because it will broke with the same error if you do it with HUGE collections and only fetching one value of the document. I think it should consume the current result and then fetch more docs, consume it and fetch more ... and so one until the `#_has_more` is false.

**Comment by jsteemann** [comment]
It sounds unnecessary to consume all the documents on the client side before iterating over them. AFAIK arangojs lazily fetches the documents so this wouldn't be required. But I maybe wrong. @pluma : can you have a look?

**Comment by pluma** [comment]
Yes, `each` fetches all results, `every` only fetches them lazily. I should adjust `each` so it behaves like `every` in that regard. I should also force `every` and `some` to cast their return values to boolean.

**Comment by pluma** [comment]
I've released the changes in 4.0.0-alpha4.
